File: service/simulator/src/simulator_resource_model.cpp
/******************************************************************
 *
 * Copyright 2015 Samsung Electronics All Rights Reserved.
 *
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 ******************************************************************/

#include "simulator_resource_model.h"
#include <sstream>
#include <boost/lexical_cast.hpp>

class to_string_visitor : public boost::static_visitor<std::string>
{
    public:
        template <typename T>
        result_type operator ()(const T &value)
        {
            return boost::lexical_cast<std::string>(value);
        }
};

class add_to_representation : public boost::static_visitor<>
{
    public:
        add_to_representation(OC::OCRepresentation &rep, const std::string &key)
            : m_rep(rep), m_key(key) {}

        template <typename T>
        void operator ()(const T &value)
        {
            m_rep.setValue(m_key, value);
        }

        OC::OCRepresentation &&getRep()
        {
            return std::move(m_rep);
        }

    private:
        OC::OCRepresentation m_rep;
        std::string m_key;
};

class range_validation : public boost::static_visitor<bool>
{
    public:
        range_validation (SimulatorResourceModel::Attribute &attrItem)
            : m_attrItem(attrItem) {}

        bool operator ()(int &value)
        {
            int min, max;
            m_attrItem.getRange(min, max);
            if (value >= min && value <= max)
                return true;
            return false;
        }

        bool operator ()(double &value)
        {
            std::vector<SimulatorResourceModel::Attribute::ValueVariant> values
                = m_attrItem.getAllowedValues();
            for (SimulatorResourceModel::Attribute::ValueVariant & val : values)
            {
                SimulatorResourceModel::Attribute::ValueVariant vVal = value;
                if (val == vVal)
                    return true;
            }
            return false;
        }

        bool operator ()(std::string &value)
        {
            std::vector<SimulatorResourceModel::Attribute::ValueVariant> values
                = m_attrItem.getAllowedValues();
            for (SimulatorResourceModel::Attribute::ValueVariant & vVal : values)
            {
                std::string val = boost::get<std::string>(vVal);
                if (val == value)
                    return true;
            }

            return false;
        }

    private:
        SimulatorResourceModel::Attribute &m_attrItem;
};

SimulatorResourceModel::Attribute::ValueVariant
&SimulatorResourceModel::Attribute::AllowedValues::at(int index)
{
    return m_values.at(index);
}

int SimulatorResourceModel::Attribute::AllowedValues::size() const
{
    return m_values.size();
}

std::string SimulatorResourceModel::Attribute::AllowedValues::toString() const
{
    std::ostringstream stream;
    stream << "[ ";
    for (auto & value : m_values)
    {
        to_string_visitor visitor;
        stream << boost::apply_visitor(visitor, value) << " ";
    }
    stream << "]";
    return stream.str();
}

std::vector<std::string> SimulatorResourceModel::Attribute::AllowedValues::toVectorString() const
{
    std::vector<std::string> values;

    for (auto & value : m_values)
    {
        to_string_visitor visitor;
        values.push_back(boost::apply_visitor(visitor, value));
    }
    return values;
}

std::vector<SimulatorResourceModel::Attribute::ValueVariant>
SimulatorResourceModel::Attribute::AllowedValues::getValues()
{
    return m_values;
}

std::string SimulatorResourceModel::Attribute::getName(void) const
{
    return m_name;
}

void SimulatorResourceModel::Attribute::setName(const std::string &name)
{
    m_name = name;
}

void SimulatorResourceModel::Attribute::getRange(int &min, int &max) const
{
    min = m_min;
    max = m_max;
}

void SimulatorResourceModel::Attribute::setRange(const int &min, const int &max)
{
    m_min = min;
    m_max = max;
}

int SimulatorResourceModel::Attribute::getAllowedValuesSize() const
{
    return m_allowedValues.size();
}

void SimulatorResourceModel::Attribute::setFromAllowedValue(const int allowedValueIndex)
{
    m_value = m_allowedValues.at(allowedValueIndex);
}

std::string SimulatorResourceModel::Attribute::valueToString() const
{
    to_string_visitor visitor;
    return boost::apply_visitor(visitor, m_value);
}

std::string SimulatorResourceModel::Attribute::allowedValuesToString() const
{
    return m_allowedValues.toString();
}

std::vector<std::string> SimulatorResourceModel::Attribute::allowedValuesToVectorString() const
{
    return m_allowedValues.toVectorString();
}

void SimulatorResourceModel::Attribute::addValuetoRepresentation(OC::OCRepresentation &rep,
        const std::string &key) const
{
    add_to_representation visitor(rep, key);
    boost::apply_visitor(visitor, m_value);
    rep = visitor.getRep();
}

bool SimulatorResourceModel::Attribute::compare(SimulatorResourceModel::Attribute &attribute)
{
    // Check the value types
    if (m_value.which() != attribute.getValue().which())
    {
        return false;
    }

    // Check the value in allowed range
    range_validation visitor(*this);
    return boost::apply_visitor(visitor, attribute.getValue());
}

std::vector<SimulatorResourceModel::Attribute::ValueVariant>
SimulatorResourceModel::Attribute::getAllowedValues()
{
    return m_allowedValues.getValues();
}

bool SimulatorResourceModel::getAttribute(const std::string &attrName, Attribute &value)
{
    if (m_attributes.end() != m_attributes.find(attrName))
    {
        value = m_attributes[attrName];
        return true;
    }

    return false;
}

std::map<std::string, SimulatorResourceModel::Attribute> SimulatorResourceModel::getAttributes()
const
{
    return m_attributes;
}

void SimulatorResourceModel::setRange(const std::string &attrName, const int min, const int max)
{
    if (m_attributes.end() != m_attributes.find(attrName))
        m_attributes[attrName].setRange(min, max);
}

void SimulatorResourceModel::setUpdateInterval(const std::string &attrName, int interval)
{
    if (m_attributes.end() != m_attributes.find(attrName))
        m_attributes[attrName].setUpdateFrequencyTime(interval);
}

void SimulatorResourceModel::updateAttributeFromAllowedValues(const std::string &attrName,
        const int allowedValueIndex)
{
    if (m_attributes.end() != m_attributes.find(attrName))
        m_attributes[attrName].setFromAllowedValue(allowedValueIndex);
}

void SimulatorResourceModel::removeAttribute(const std::string &attrName)
{
    m_attributes.erase(attrName);
    return;
}

OC::OCRepresentation SimulatorResourceModel::getOCRepresentation() const
{
    OC::OCRepresentation rep;
    for (auto & attribute : m_attributes)
    {
        (attribute.second).addValuetoRepresentation(rep, attribute.first);
    }

    return rep;
}

bool SimulatorResourceModel::update(OC::OCRepresentation &ocRep)
{
    if (0 == ocRep.size())
        return true;

    // Convert OCRepresentation to SimulatorResourceModel
    SimulatorResourceModel resModel = create(ocRep);

    return update(resModel);
}

bool SimulatorResourceModel::update(SimulatorResourceModel &repModel)
{
    std::map<std::string, SimulatorResourceModel::Attribute> attributes = repModel.getAttributes();
    for (auto & attributeItem : attributes)
    {
        // Check the attribute presence
        SimulatorResourceModel::Attribute attribute;
        if (false == getAttribute((attributeItem.second).getName(), attribute))
        {
            return false;
        }

        // Check the validity of the value to be set
        if (false == attribute.compare(attributeItem.second))
        {
            return false;
        }
        m_attributes[(attributeItem.second).getName()].setValue((attributeItem.second).getValue());
    }

    return true;
}

SimulatorResourceModel SimulatorResourceModel::create(const OC::OCRepresentation &ocRep)
{
    SimulatorResourceModel resModel;
    for (auto & attributeItem : ocRep)
    {
        SimulatorResourceModel::Attribute attribute;
        if (attributeItem.type() == OC::AttributeType::Integer)
            attribute.setValue(attributeItem.getValue<int>());
        if (attributeItem.type() == OC::AttributeType::Double)
            attribute.setValue(attributeItem.getValue<double>());
        if (attributeItem.type() == OC::AttributeType::String)
            attribute.setValue(attributeItem.getValue<std::string>());

        attribute.setName(attributeItem.attrname());
        resModel.m_attributes[attributeItem.attrname()] = attribute;
    }

    return resModel;
}

File: service/simulator/java/sdk/src/org/oic/simulator/SimulatorResourceModel.java
/*
 * Copyright 2015 Samsung Electronics All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.oic.simulator;

import java.util.HashMap;
import java.util.Map;

/**
 * This class represents the resource model of a resource. A resource model
 * contains a set of attributes and their properties. It represents the complex
 * value type of an attribute.
 */
public class SimulatorResourceModel {

    private Map<String, AttributeValue>    mValues     = null;
    private Map<String, AttributeProperty> mProperties = null;

    /**
     * Constructs new {@link SimulatorResourceModel} object.
     */
    public SimulatorResourceModel() {
        mValues = new HashMap<>();
        mProperties = new HashMap<>();
    }

    /**
     * API to add an attribute to resource model.
     *
     * @param attrName
     *            Name of the attribute.
     * @param value
     *            Value of the attribute.
     *
     * @throws InvalidArgsException
     *             This exception will be thrown on invalid input.
     */
    public void addAttribute(String attrName, AttributeValue value)
            throws InvalidArgsException {
        if (null == attrName || attrName.isEmpty())
            throw new InvalidArgsException("Invalid attribute name!");

        if (null == value)
            throw new InvalidArgsException("Attribute value is null!");

        mValues.put(attrName, value);
    }

    /**
     * API to add an attribute to resource model.
     *
     * @param attribute
     *            {@link SimulatorResourceAttribute} to be added to resource
     *            model.
     *
     * @throws InvalidArgsException
     *             This exception will be thrown on invalid input.
     */
    public void addAttribute(SimulatorResourceAttribute attribute)
            throws InvalidArgsException {
        if (null == attribute || null == attribute.name()
                || attribute.name().isEmpty() || null == attribute.value())
            throw new InvalidArgsException("Invalid attribute!");

        mValues.put(attribute.name(), attribute.value());
    }

    /**
     * API to set attribute's property.
     *
     * @param attrName
     *            Name of the attribute.
     * @param property
     *            {@link AttributeProperty} to be set for attribute.
     *
     * @throws InvalidArgsException
     *             This exception will be thrown on invalid input.
     */
    public void setAttributeProperty(String attrName, AttributeProperty property)
            throws InvalidArgsException {
        if (null == attrName || attrName.isEmpty())
            throw new InvalidArgsException("Invalid attribute!");

        if (null == property)
            throw new InvalidArgsException("Invalid attribute property!");

        mProperties.put(attrName, property);
    }

    /**
     * API to set attribute's value.
     *
     * @param attrName
     *            Name of the attribute.
     * @param value
     *            {@link AttributeValue} to be set for attribute.
     *
     * @throws InvalidArgsException
     *             This exception will be thrown on invalid input.
     */
    public void setAttributeValue(String attrName, AttributeValue value)
            throws InvalidArgsException {
        if (null == attrName || attrName.isEmpty())
            throw new InvalidArgsException("Invalid attribute name!");

        if (null == value)
            throw new InvalidArgsException("Attribute value is null!");

        mValues.put(attrName, value);
    }

    /**
     * API to get all the attributes of resource model.
     *
     * @return Map of attributes with attribute name as key and its
     *         corresponding {@link SimulatorResourceAttribute} as value.
     */
    public Map<String, SimulatorResourceAttribute> getAttributes() {
        Map<String, SimulatorResourceAttribute> attributes = new HashMap<>();
        for (Map.Entry<String, AttributeValue> entry : mValues.entrySet()) {
            SimulatorResourceAttribute attribute = new SimulatorResourceAttribute(
                    entry.getKey(), entry.getValue(), mProperties.get(entry
                            .getKey()));
            attributes.put(entry.getKey(), attribute);
        }

        return attributes;
    }

    /**
     * API to get attribute by name.
     *
     * @param attrName
     *            Name of the attribute.
     *
     * @return {@link SimulatorResourceAttribute}.
     */
    public SimulatorResourceAttribute getAttribute(String attrName) {
        if (mValues.containsKey(attrName)) {
            SimulatorResourceAttribute attribute = new SimulatorResourceAttribute(
                    attrName, mValues.get(attrName), mProperties.get(attrName));
            return attribute;
        }

        return null;
    }

    /**
     * API to check whether resource model has an attribute with given name.
     *
     * @param attrName
     *            Name of the attribute.
     *
     * @return true if resource model has an attribute with given name,
     *         otherwise false.
     */
    public boolean containsAttribute(String attrName) {
        if (mValues.containsKey(attrName))
            return true;
        return false;
    }

    /**
     * API to get value type information of attribute.
     *
     * @param attrName
     *            Name of the attribute.
     *
     * @return Attribute value type information {@link AttributeValue.TypeInfo
     * 
     * }.
     */
    public AttributeValue.TypeInfo getAttributeType(String attrName) {
        if (mValues.containsKey(attrName))
            return mValues.get(attrName).typeInfo();
        return null;
    }

    /**
     * API to remove attribute from resource model.
     *
     * @param attrName
     *            Name of the attribute.
     *
     * @return true if resource model has an attribute with the given name,
     *         otherwise false.
     */
    public boolean removeAttribute(String attrName) {
        if (mValues.containsKey(attrName)) {
            mValues.remove(attrName);
            if (mProperties.containsKey(attrName))
                mProperties.remove(attrName);
            return true;
        }

        return false;
    }

    /**
     * API to get number of attributes present in resource model.
     *
     * @return Number of attributes present in resource model.
     */
    public int size() {
        return mValues.size();
    }

    /**
     * API to update the attribute values from given
     * {@link SimulatorResourceModel}.
     * 
     * @param resourceModel
     *            {@link SimulatorResourceModel} holding the new attribute
     *            values.
     */
    public void update(SimulatorResourceModel resourceModel) {
        if (null == resourceModel || 0 == resourceModel.size())
            return;

        for (Map.Entry<String, SimulatorResourceAttribute> entry : resourceModel
                .getAttributes().entrySet()) {
            SimulatorResourceAttribute newAttribute = entry.getValue();
            SimulatorResourceAttribute attribute = getAttribute(entry.getKey());
            if (null != newAttribute && null != attribute) {
                if (null != attribute.property()) {
                    AttributeValueValidation validation = new AttributeValueValidation(
                            attribute.property());
                    if (!validation.validate(newAttribute.value())) {
                        mValues.put(entry.getKey(), newAttribute.value());
                    }
                } else {
                    mValues.put(entry.getKey(), newAttribute.value());
                }
            }
        }
    }

    private SimulatorResourceModel(Map<String, AttributeValue> values,
            Map<String, AttributeProperty> properties) {
        mValues = values;
        mProperties = properties;
    }
}
